src/source/Ice/Tls.c

```diff
@@ -0,0 +1,267 @@
+    tlsSessionChangeState(pTlsSession, TLS_SESSION_STATE_CONNECTING);
+    SSL_do_handshake(pTlsSession->pSsl);
+
+    /* send handshake */
+    CHK_STATUS(tlsSessionPutApplicationData(pTlsSession, NULL, 0));
+
+CleanUp:
+
+    CHK_LOG_ERR(retStatus);
+
+    if (STATUS_FAILED(retStatus)) {
+        if (freeBios) {
+            if (pReadBio != NULL) {
+                BIO_free(pReadBio);
+            }
+
+            if (pWriteBio != NULL) {
+                BIO_free(pWriteBio);
+            }
+        }
+        ERR_print_errors_fp (stderr);
+    }
+
+    LEAVES();
+    return retStatus;
+}
+
+STATUS tlsSessionProcessPacket(PTlsSession pTlsSession, PBYTE pData, UINT32 bufferLen, PUINT32 pDataLen)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    BOOL continueRead = TRUE;
+    INT32 sslReadRet = 0;
+    UINT32 writtenBytes = 0;
+    UINT64 sslErrorRet;
+
+    CHK(pTlsSession != NULL && pData != NULL && pDataLen != NULL, STATUS_NULL_ARG);
+    CHK(pTlsSession->state != TLS_SESSION_STATE_NEW, STATUS_SOCKET_CONNECTION_NOT_READY_TO_SEND);
+    CHK(pTlsSession->state != TLS_SESSION_STATE_CLOSED, STATUS_SOCKET_CONNECTION_CLOSED_ALREADY);
+
+    // return early if there's no data
+    CHK(*pDataLen > 0, retStatus);
+
+    CHK(BIO_write(SSL_get_rbio(pTlsSession->pSsl), pData, *pDataLen) > 0, STATUS_SECURE_SOCKET_READ_FAILED);
+
+    // read as much as possible
+    while(continueRead && writtenBytes < bufferLen) {
+        sslReadRet = SSL_read(pTlsSession->pSsl, pData + writtenBytes, bufferLen - writtenBytes);
+        if (sslReadRet <= 0) {
+            sslReadRet = SSL_get_error(pTlsSession->pSsl, sslReadRet);
+            switch (sslReadRet) {
+                case SSL_ERROR_WANT_WRITE:
+                    continueRead = FALSE;
+                    break;
+                case SSL_ERROR_WANT_READ:
+                    break;
+                default:
+                    sslErrorRet = ERR_get_error();
+                    DLOGW("SSL_read failed with %s", ERR_error_string(sslErrorRet, NULL));
+                    break;
+            }
+            break;
+        } else {
+            writtenBytes += sslReadRet;
+        }
+    }
+
+    *pDataLen = writtenBytes;
+
+CleanUp:
+
+    // CHK_LOG_ERR might be too verbose
+    if (STATUS_FAILED(retStatus)) {
+        DLOGD("Warning: reading socket data failed with 0x%08x", retStatus);
+    }
+
+    return retStatus;
+}
+
+STATUS tlsSessionPutApplicationData(PTlsSession pTlsSession, PBYTE pData, UINT32 dataLen)
+{
+    STATUS retStatus = STATUS_SUCCESS;
+    INT32 sslRet = 0, sslErr = 0;
+
+    SIZE_T wBioDataLen = 0;
+    PCHAR wBioBuffer = NULL;
+
+    CHK(pTlsSession != NULL, STATUS_NULL_ARG);
+
+    if (SSL_is_init_finished(pTlsSession->pSsl)) {
+        tlsSessionChangeState(pTlsSession, TLS_SESSION_STATE_CONNECTED);
+
+        /* Should have a valid buffer */
+        CHK(pData != NULL && dataLen > 0, STATUS_INVALID_ARG);
+        sslRet = SSL_write(pTlsSession->pSsl, pData, dataLen);
```

Comment:
can pData be NULL here? If not, make sure you either have validated pData in above call sites or validate here with pTlsSession not being NULL

Reply:
Yep, makes sense. pData shouldn't be null here, since it's already checked in the above call.